PyDynamic/model_estimation/fit_filter.py

```diff
@@ -87,92 +86,207 @@ def _fitIIR(
+                print(
+                    f"LSIIR: Fitting "
+                    f"{'' if UHvals is None else f'for MC run {mc_run} '}"
+                    f"finished. No stabilization requested"
+                    f". {warning_unstable if unstable else ''} "
+                    f"Final sum of squares = {sos}"
+                )
+            # Since no stabilization was done, we return tau = 0 regardless of the
+            # desired initial estimate of the time delay provided for the stabilization.
+            taus[mc_run] = 0
+            continue
         else:
-            stable = True
-        run = run + 1
-
-    if np.count_nonzero(np.abs(np.roots(a)) > 1) > 0:
-        print("Caution: The algorithm did NOT result in a stable IIR filter!")
-        print(
-            "Maybe try again with a higher value of tau0 or a higher filter "
-            "order?")
+            # Otherwise we initialize the time delay for the current Monte Carlo run.
+            taus[mc_run] = tau
+
+        # Initialize counter which we use to report about required iteration count.
+        current_stab_iter = 0
+        # Stabilize filter coefficients with a maximum number of iterations.
+        while unstable and current_stab_iter < max_stab_iter:
+            # Compute appropriate time delay for the stabilization of the filter.
+            a_stab = mapinside(a_i)
+            g_1 = grpdelay(b_i, a_i, Fs)[0]
+            g_2 = grpdelay(b_i, a_stab, Fs)[0]
+            taus[mc_run] += np.ceil(np.median(g_2 - g_1))
+
+            # Conduct stabilization step through time delay.
+            b_i, a_i = _fitIIR(Hvals, taus[mc_run], w, E, Na, Nb, inv=inv)
+
+            # Prepare abortion in case filter is stable.
+            unstable = not isstable(b=b_i, a=a_i, ftype="digital")
+            current_stab_iter += 1
+
+        # Finally store filter stacked filter parameters.
+        as_and_bs[mc_run, :] = np.hstack((a_i[1:], b_i))
+        stab_iters[mc_run] = current_stab_iter
+
+    b = np.mean(as_and_bs[:, Na:], axis=0)
+    a = np.hstack((np.array([1.0]), np.mean(as_and_bs[:, :Na], axis=0)))
+    tau = np.mean(taus)
+    stab_iter = np.mean(stab_iters)
 
-    print(
-        "Least squares fit finished after %d iterations (tau=%d)." % (run, tau))
-    Hd = dsp.freqz(b, a, 2 * np.pi * f / Fs)[1]
-    Hd = Hd * np.exp(1j * 2 * np.pi * f / Fs * tau)
-    res = np.hstack(
-        (np.real(Hd) - np.real(Hvals), np.imag(Hd) - np.imag(Hvals)))
-    rms = np.sqrt(np.sum(res ** 2) / len(f))
-    print("Final rms error = %e \n\n" % rms)
+    if verbose:
+        if unstable:
+            print(
+                f"LSIIR: {warning_unstable} Maybe try again with a higher value of "
+                f"tau or a higher filter order? Least squares fit finished after "
+                f"{stab_iter} stabilization iterations "
+                f"{'on average ' if UHvals and mc_runs > 1 else ''}(tau = {tau})."
+            )
```

Comment:
This message should reflect the new algorithm in which we drop the unstable filters from the computation of the final result.

Reply:
This is resolved in 43ebfd5